### PR TITLE
Implements if, for, return, let AND lambdas for Noodle.

### PR DESCRIPTION
**Pull request by andyHa**
* **if**: `let x = 5; if (3 < 4) { x = 3; } else { x = 4; }; return x;`
* **let**: `let x = 3; x = 4; return x;`
* **for**: `let sum = 0; for(int x : java.util.Arrays.asList(3, 4)) { sum = sum + x; }; return sum;`
* **lambda**: `let sum = 0; java.util.Arrays.asList(3,4).forEach(|x| sum = sum + x); return sum;`

### CODE REVIEW

File: src/test/java/sirius/pasta/noodle/compiler/CompilerSpec.groovy [hunk 1]
```diff
@@ -37,15 +47,43 @@ class CompilerSpec extends BaseSpecification {
...
+        and:
+        compile("let x = 3; x = 4; return x;").call(new SimpleEnvironment()) == 4
+        and:
+        compile("let sum = 0; for(int x : java.util.Arrays.asList(3, 4)) { sum = sum + x; }; return sum;").
+                call(new SimpleEnvironment()) == 7
+    }
+
+    def "parsing lambdas works"() {
+        expect: "Simple generic type propagation works"
+        compile("let sum = 0; NoodleExample.intStream().forEach(|x| sum = sum + x); return sum;").
```
Comment (qw3ry) [inline src/test/java/sirius/pasta/noodle/compiler/CompilerSpec.groovy:76]: Gibts nen speziellen Grund für die `|x|`-Syntax? In Java würde man ja eher `x ->` erwarten...

Comment (andyHa) [reply src/test/java/sirius/pasta/noodle/compiler/CompilerSpec.groovy:76]: yep -> is way harder to parse, as it needs an unbound lookahead (we need to consume all parameters commas and braces to eventually check if there is a -> or not - '|' on the other hand can be checked with a lookup of 1 :) - we can still add lambda syntax support later if really needed...

Comment (idlira) [reply src/test/java/sirius/pasta/noodle/compiler/CompilerSpec.groovy:76]: wie wäre es mit einen JS `function(x) sum = sum + x` oder Python `lambda x : sum = sum + x` style?

Comment (andyHa) [reply src/test/java/sirius/pasta/noodle/compiler/CompilerSpec.groovy:76]: The mother of all languages Smalltalk uses this syntax :-P - and rust and a whole bunch of others do as well...

**Review by idlira — APPROVED** [review]

**Review by meg1502 — APPROVED** [review]

File: src/main/java/sirius/pasta/noodle/compiler/ir/ForStatement.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import sirius.pasta.noodle.compiler.Assembler;
+import sirius.pasta.noodle.compiler.CompilationContext;
+import sirius.pasta.noodle.compiler.VariableScoper;
+
+import javax.annotation.Nullable;
+import java.lang.reflect.Type;
+import java.util.Iterator;
+
+/**
+ * Represents an IF statement.
```
Comment (sabieber) [inline src/main/java/sirius/pasta/noodle/compiler/ir/ForStatement.java:24]: IF -> FOR

File: src/main/java/sirius/pasta/noodle/compiler/TypeTools.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.lang.reflect.WildcardType;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * A helper which tries to resolve all type variables of a class (and method) into concrete types.
```
Comment (sabieber) [inline src/main/java/sirius/pasta/noodle/compiler/TypeTools.java:23]: `Tries to ...` or `Provides a helper` (begin JavaDoc with a verb)

File: src/main/java/sirius/pasta/noodle/LambdaEnvironment.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ * http://www.scireum.de - info@scireum.de
+ */
+
+package sirius.pasta.noodle;
+
+/**
+ * Provides a wrapper environment for a ({@link sirius.pasta.noodle.compiler.ir.LambdaNode lambda}).
+ * <p>
+ * This essentially shadows the few environment variables which are local to the lambda and stores them
+ * in a local buffer. Everything else if passed through to the enclosing environment.
```
Comment (sabieber) [inline src/main/java/sirius/pasta/noodle/LambdaEnvironment.java:15]: if -> is

**Review by sabieber — CHANGES_REQUESTED** [review]

**Review by mkeckmkeck — CHANGES_REQUESTED** [review]
typos

File: src/main/java/sirius/pasta/noodle/LambdaEnvironment.java (hunk 4 above)
Comment (andyHa) [inline src/main/java/sirius/pasta/noodle/LambdaEnvironment.java:15]:
```suggestion
 * in a local buffer. Everything else is passed through to the enclosing environment.
```

File: src/main/java/sirius/pasta/noodle/compiler/TypeTools.java (hunk 3 above)
Comment (andyHa) [inline src/main/java/sirius/pasta/noodle/compiler/TypeTools.java:23]:
```suggestion
 * Provides a helper which tries to resolve all type variables of a class (and method) into concrete types.
```

**Review by sabieber — APPROVED** [review]

**Review by idlira — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Review by jmuscireum — APPROVED** [review]

**Review by qw3ry — APPROVED** [review]

**Review by mkeckmkeck — APPROVED** [review]

**Pull request merged**